### PR TITLE
Refactor build for local, staging, and prod builds

### PR DESCRIPTION
**Pull request by awong-dev**
== Background ==
There are 4 effective environments.
 (a) Local desktop for iterative development.
 (b) Production (must NOT have unlreleased dev features)
 (c) Staging (must HAVE unreleased dev features)
 (d) CI PR builds which must match the environment they are merging to.

There are 3 dimensions in the above:
 (1) JS Optimization flags. On for b, c, and d but off for a.
 (2) Unreleased "dev" features. On for a, c and off for b.
     Conditionally off for d.
 (3) Output directory. _site for all but b. _site_production for b.

3 is a defensive safeguard against a dev build accidentally getting
pushed to prod (has happened before). These can be collapsed into 2
flags "enable_dev_features" and "enable_optimizations".

== Change summary ==
This CL adds rake targets that support builds varying in the reduced two
dimensions. Summary of changes to targets are as follows:
- Removed: build_all_configurations
  This was a hack used previously to make Travis always build _site and
  _site_production. However, it caused a race condition on production
  push due to the use of `multitask`. Is is no longer needed so removing.
- Changed: build
  This target now inspects the environment and either builds with
  development features on or off based on whether or not it sees a
  production branch Travis build.
- Changed: build_production -> build_prod
  Renamed to be consistent with "prod" naming used elsewhere.
  This target builds into _site_production with development features
  disabled. It has optimizations enabled if building in CI.
- Changed: build_dev
  This target builds into _site with development features enabled.
  It has optimizations enabled if building in CI.
- No change: webpack_watch
  Always builds without optimizations with dev features on.
- Added: ci_dev
  Convenience target for simulating a ci dev BUILD locally.
- Added: ci_prod
  Convenience target for simulating a ci prod BUILD locally.

Other changes:
- jekyll `site.production` variable configured correctly. Previously
  it was set but unused.
- Development sourcemaps changed to cheap version for faster
  builds.
- Javascript bundle now has environment name mangled into the path.
  Again, defensive coding to avoid crossing wires on deployment.

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -1,5 +1,10 @@
 require "tmpdir"
 
+# TODO(awong): Add method for overriding these if building on the local
+# machine. Currently it keys off of Travis configs only which makes it
+# less natural as an "API" for local development.
+enable_dev_features = (ENV['TRAVIS_BRANCH'] == "production")
```
Comment (ayaleloehr) [inline Rakefile]: Looks to me like the logic here is "enable the development features if the branch being merged to is production"  But don't we only want development features when it is **not** production?

File: Rakefile [hunk 2]
```diff
@@ -31,24 +36,30 @@ task :bootstrap do
   sh "brew install npm"
 
   # Continue with install.
-  Rake::Task["install"].invoke
+  Rake::Task[:install].invoke
 end
 
-desc "Build the website"
+desc "Build the website with development features"
```
Comment (ayaleloehr) [inline Rakefile]: This description feels inaccurate, as it says `build with development features`, but the task runs a build either with or without development features, depending on how environment variables are set. Maybe change it back to just `Build the website`?

File: Rakefile [hunk 3]
```diff
@@ -77,19 +88,29 @@ task :lint do
   sh "npm run-script lint"
 end
 
-desc "Run webpack with development settings"
-task :webpack do
-  sh "BUILD_TYPE=dev npm run-script webpack"
+desc "Run webpack with development features turned on. Build is optimized."
```
Comment (ayaleloehr) [inline Rakefile]: Remove `Build is optimized.`, as there is a conditional in this task to build whether or not `enable_optimizations` is set.

**Comment by ayaleloehr** [comment]
@awong-dev looks like we went to a naming convention of `prod` instead of `production` for all of the Rake file, but that many of the other files are using the full word, `production`.  Examples of this include `_config_production.yml`, `webpack_production.config.js`, and the production code is deployed to `./_site_production`.  Do you think we need to standardize `prod` vs `production` across these files as well?

File: Rakefile [hunk 4]
```diff
@@ -77,19 +88,29 @@ task :lint do
...
+task :webpack_dev do
+  sh "rm -rf assets/js/generated"
+  if enable_optimizations then
+    sh "BUILD_TYPE=dev npm run-script webpack"
+  else
+    sh "BUILD_TYPE=dev npm run-script webpack-noopt"
```
Comment (ayaleloehr) [inline Rakefile]: Assuming `noopt` stands for `no-optimization`?  If this is the case, why is it `noopt` instead of `no-opt`?

**Comment by ayaleloehr** [comment]
@awong-dev liking this overall!  Most of my comments are nit-picky naming things, but since this is deployment, which always feels like voodoo anyways, getting those details consistent will hopefully make life easier when something is broken and we're stressed about understanding and fixing it quickly.

File: Rakefile (hunk 1 above)
Comment (awong-dev) [inline Rakefile]: ick. Good catch. I fixed the usage.

File: Rakefile (hunk 2 above)
Comment (awong-dev) [inline Rakefile]: Reworded.

File: Rakefile (hunk 3 above)
Comment (awong-dev) [inline Rakefile]: done.

File: Rakefile (hunk 4 above)
Comment (awong-dev) [inline Rakefile]: Too many dashes :(

Also, I was thinking of each dash section as being a logical modifier.

**Comment by awong-dev** [comment]
Agreed about making prod/production standardized, but would like to punt to a followup PR since this one has enough churn as is.

**Comment by awong-dev** [comment]
@ayaleloehr Thanks for the review!  You say "nit-picky" and I say "correct level of quality for any review" :)

PTAL.

**Comment by ayaleloehr** [comment]
LGTM.  Says branch is out of date with base though. Can you fix and merge @awong-dev ?

**Comment by awong-dev** [comment]
Will do. Thank you!

**Pull request merged**